src/provider.rs

```diff
@@ -174,6 +197,164 @@ impl TableProvider for SeafowlTable {
+        data_type: &DataType,
+    ) -> Result<ScalarValue> {
+        match bytes_value.as_ref() {
+            Some(bytes) => match String::from_utf8(bytes.clone()) {
+                Ok(string_val) => ScalarValue::try_from_string(string_val, data_type),
+                Err(error) => Err(DataFusionError::Internal(format!(
+                    "Failed to parse min/max value: {}",
+                    error
+                ))),
+            },
+            None => Ok(ScalarValue::Utf8(None)),
+        }
+    }
+
+    // Prune away partitions that are refuted by the provided filter expressions
+    async fn prune(&self, filters: &[Expr]) -> Vec<SeafowlPartition> {
+        let mut partitions = self.partitions.deref().clone();
+
+        if filters.is_empty() {
+            // No qualifiers means we need all partitions
+            return partitions;
+        }
+
+        let mut partition_mask = vec![false; self.partition_count];
+
+        for expr in filters {
+            match PruningPredicate::try_new(expr.clone(), self.schema.clone()) {
+                Ok(predicate) => match predicate.prune(self) {
+                    Ok(expr_mask) => {
+                        partition_mask = partition_mask
+                            .iter()
+                            .zip(expr_mask)
+                            .map(|(&a, b)| a || b)
+                            .collect()
+                    }
+                    Err(error) => {
+                        warn!(
+                            "Failed pruning the partitions for expr {:?}: {}",
+                            expr, error
+                        );
+                        return partitions;
+                    }
+                },
+                Err(error) => {
+                    warn!(
+                        "Failed constructing a pruning predicate for expr {:?}: {}",
+                        expr, error
+                    );
+                    return partitions;
+                }
+            }
+        }
+
+        let mut iter = partition_mask.iter();
+        partitions.retain(|_| *iter.next().unwrap());
+        partitions
+    }
+
+    // Try to convert the vector of min/max scalar values for a column into a generic array reference
+    fn get_values_array(&self, values: Option<&Vec<ScalarValue>>) -> Option<ArrayRef> {
+        match values {
+            Some(stats) => {
+                match ScalarValue::iter_to_array(stats.clone().into_iter()) {
+                    Ok(array) => Some(array),
+                    Err(error) => {
+                        warn!("Failed to convert vector of min/max values {:?} to array: {}", values, error);
+                        None
+                    }
+                }
+            }
+            None => None,
+        }
+    }
+}
+
+impl PruningStatistics for SeafowlPruningStatistics {
+    fn min_values(&self, column: &Column) -> Option<ArrayRef> {
+        self.get_values_array(self.min_values.get(column.name.as_str()))
```

Comment:
Why not store the `min_values`/`max_values`/`null_counts` as `Option<ArrayRef>`s to begin with? i.e.: `min_values: HashMap<String, Option<ArrayRef>>`? You're converting them into a Vec anyway in the constructor (and cloning in `get_values_array`, so I don't think this will result in a worse performance.

Reply:
I tried that initially, but it was a pita to get it working; the issue is that building `ArrayRef` incrementally (like we do it in `from_partitions` by iterating over partitions/columns) is hard/unsupported, and only this conversion of the final vector via [`ScalarValue::iter_to_array`](https://github.com/apache/arrow-datafusion/blob/e0a9fa347f7adb1e6d6157b537e44c574e88244a/datafusion/common/src/scalar.rs#L806) worked.

Reply:
This is doable off course, but I was thinking that going for the opposite, JIT-like, approach may be even more beneficial: don't initialize `min_values`/`max_values`/`null_counts` at all, and only upon a call to the `PruningStatistics` interface do a get-or-insert map lookup op for that particular column's requested stats. 

Since the number of columns in a table is usually much larger than the number of distinct columns used in query qualifiers I expect this to be the most performant approach.